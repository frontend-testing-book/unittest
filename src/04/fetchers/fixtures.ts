import type { Article, Articles, HttpError } from './type'

export const httpError: HttpError = {
  err: { message: 'internal server error' },
}

export const getMyArticlesData: Articles = {
  articles: [
    {
      id: 'howto-testing-with-typescript',
      createdAt: '2022-07-19T22:38:41.005Z',
      tags: ['testing'],
      title: 'TypeScript を使ったテストの書き方',
      body: 'テストを書く時、TypeScript を使うことで、テストの保守性が向上します…',
    },
    {
      id: 'nextjs-link-component',
      createdAt: '2022-07-19T22:38:41.005Z',
      tags: ['nextjs'],
      title: 'Next.js の Link コンポーネント',
      body: 'Next.js の画面遷移には、Link コンポーネントを使用します…',
    },
    {
      id: 'react-component-testing-with-jest',
      createdAt: '2022-07-19T22:38:41.005Z',
      tags: ['testing', 'react'],
      title: 'Jest ではじめる React のコンポーネントテスト',
      body: 'Jest は単体テストとして、UIコンポーネントのテストが可能です…',
    },
  ],
}

export const postMyArticleData: Article = {
  id: 'xxxxxxx-123456',
  createdAt: '2022-07-19T22:38:41.005Z',
  tags: ['testing', 'react'],
  title: 'Jest ではじめる React のコンポーネントテスト',
  body: 'Jest は単体テストとして、UIコンポーネントのテストが可能です。',
}
