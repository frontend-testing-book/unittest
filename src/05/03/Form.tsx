type Props = {
  name: string
  onSubmit?: (event: React.FormEvent<HTMLFormElement>) => void
}
export const Form = ({ name, onSubmit }: Props) => {
  return (
    <form
      onSubmit={(event) => {
        event.preventDefault()
        onSubmit?.(event)
      }}
    >
      <h2>アカウント情報</h2>
      <p>{name}</p>
      <div>
        <button>編集する</button>
      </div>
    </form>
  )
}
