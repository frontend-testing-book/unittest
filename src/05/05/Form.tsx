import { useId, useState } from 'react'

import { Agreement } from './Agreement'
import { InputAccount } from './InputAccount'

export const Form = () => {
  const [checked, setChecked] = useState(false)
  const headingId = useId()
  return (
    <form aria-labelledby={headingId}>
      <h2 id={headingId}>新規アカウント登録</h2>
      <InputAccount />
      <Agreement
        onChange={(event) => {
          setChecked(event.currentTarget.checked)
        }}
      />
      <div>
        <button disabled={!checked}>サインアップ</button>
      </div>
    </form>
  )
}
