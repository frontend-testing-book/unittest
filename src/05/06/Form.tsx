import { useState } from 'react'

import { ContactNumber } from './ContactNumber'
import { DeliveryAddress } from './DeliveryAddress'
import { PastDeliveryAddress } from './PastDeliveryAddress'
import { RegisterDeliveryAddress } from './RegisterDeliveryAddress'

export type AddressOption = React.ComponentProps<'option'> & { id: string }
export type Props = {
  deliveryAddresses?: AddressOption[]
  onSubmit?: (event: React.FormEvent<HTMLFormElement>) => void
}
export const Form = (props: Props) => {
  const [registerNew, setRegisterNew] = useState<boolean | undefined>(undefined)
  return (
    <form onSubmit={props.onSubmit}>
      <h2>お届け先情報の入力</h2>
      <ContactNumber />
      {props.deliveryAddresses?.length ? (
        <>
          <RegisterDeliveryAddress onChange={setRegisterNew} />
          {registerNew ? (
            <DeliveryAddress title="新しいお届け先" />
          ) : (
            <PastDeliveryAddress
              disabled={registerNew === undefined}
              options={props.deliveryAddresses}
            />
          )}
        </>
      ) : (
        <DeliveryAddress />
      )}
      <hr />
      <div>
        <button>注文内容の確認へ進む</button>
      </div>
    </form>
  )
}
