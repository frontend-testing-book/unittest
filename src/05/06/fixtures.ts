export const deliveryAddresses = [
  {
    id: 'address_id_xxxx',
    value: 'address_id_xxxx',
    children: '〒167-0051 東京都杉並区荻窪1-00-00',
  },
]
