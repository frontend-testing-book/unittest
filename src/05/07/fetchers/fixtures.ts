import type { HttpError, Result } from './type'

export const httpError: HttpError = {
  err: { message: 'internal server error' },
}

export const postMyAddressMock: Result = {
  result: 'ok',
}
