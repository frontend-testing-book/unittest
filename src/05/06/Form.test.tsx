import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { deliveryAddresses } from './fixtures'
import { Form } from './Form'

const user = userEvent.setup()

async function inputContactNumber(
  inputValues = {
    name: '田中 太郎',
    phoneNumber: '000-0000-0000',
  },
) {
  await user.type(
    screen.getByRole('textbox', { name: '電話番号' }),
    inputValues.phoneNumber,
  )
  await user.type(
    screen.getByRole('textbox', { name: 'お名前' }),
    inputValues.name,
  )
  return inputValues
}

async function inputDeliveryAddress(
  inputValues = {
    postalCode: '167-0051',
    prefectures: '東京都',
    municipalities: '杉並区荻窪1',
    streetNumber: '00-00',
  },
) {
  await user.type(
    screen.getByRole('textbox', { name: '郵便番号' }),
    inputValues.postalCode,
  )
  await user.type(
    screen.getByRole('textbox', { name: '都道府県' }),
    inputValues.prefectures,
  )
  await user.type(
    screen.getByRole('textbox', { name: '市区町村' }),
    inputValues.municipalities,
  )
  await user.type(
    screen.getByRole('textbox', { name: '番地番号' }),
    inputValues.streetNumber,
  )
  return inputValues
}

async function clickSubmit() {
  await user.click(screen.getByRole('button', { name: '注文内容の確認へ進む' }))
}

function mockHandleSubmit() {
  const mockFn = jest.fn()
  const onSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault()
    const formData = new FormData(event.currentTarget)
    const data: { [k: string]: unknown } = {}
    formData.forEach((value, key) => (data[key] = value))
    mockFn(data)
  }
  return [mockFn, onSubmit] as const
}

describe('過去のお届け先がない場合', () => {
  test('お届け先入力欄がある', () => {
    render(<Form />)
    expect(screen.getByRole('group', { name: '連絡先' })).toBeInTheDocument()
    expect(screen.getByRole('group', { name: 'お届け先' })).toBeInTheDocument()
  })

  test('入力・送信すると、入力内容が送信される', async () => {
    const [mockFn, onSubmit] = mockHandleSubmit()
    render(<Form onSubmit={onSubmit} />)
    const contactNumber = await inputContactNumber()
    const deliveryAddress = await inputDeliveryAddress()
    await clickSubmit()
    expect(mockFn).toHaveBeenCalledWith(
      expect.objectContaining({ ...contactNumber, ...deliveryAddress }),
    )
  })

  test('Snapshot', () => {
    const { container } = render(<Form />)
    expect(container).toMatchSnapshot()
  })
})

describe('過去のお届け先がある場合', () => {
  test('設問に答えるまで、お届け先を選べない', () => {
    render(<Form deliveryAddresses={deliveryAddresses} />)
    expect(
      screen.getByRole('group', { name: '新しいお届け先を登録しますか？' }),
    ).toBeInTheDocument()
    expect(screen.getByRole('group', { name: '過去のお届け先' })).toBeDisabled()
  })

  test('「いいえ」を選択・入力・送信すると、入力内容が送信される', async () => {
    const [mockFn, onSubmit] = mockHandleSubmit()
    render(<Form deliveryAddresses={deliveryAddresses} onSubmit={onSubmit} />)
    await user.click(screen.getByLabelText('いいえ'))
    expect(
      screen.getByRole('group', { name: '過去のお届け先' }),
    ).toBeInTheDocument()
    const inputValues = await inputContactNumber()
    await clickSubmit()
    expect(mockFn).toHaveBeenCalledWith(expect.objectContaining(inputValues))
  })

  test('「はい」を選択・入力・送信すると、入力内容が送信される', async () => {
    const [mockFn, onSubmit] = mockHandleSubmit()
    render(<Form deliveryAddresses={deliveryAddresses} onSubmit={onSubmit} />)
    await user.click(screen.getByLabelText('はい'))
    expect(
      screen.getByRole('group', { name: '新しいお届け先' }),
    ).toBeInTheDocument()
    const contactNumber = await inputContactNumber()
    const deliveryAddress = await inputDeliveryAddress()
    await clickSubmit()
    expect(mockFn).toHaveBeenCalledWith(
      expect.objectContaining({ ...contactNumber, ...deliveryAddress }),
    )
  })

  test('Snapshot', () => {
    const { container } = render(<Form deliveryAddresses={deliveryAddresses} />)
    expect(container).toMatchSnapshot()
  })
})
