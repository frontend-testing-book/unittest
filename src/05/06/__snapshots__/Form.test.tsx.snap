// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`過去のお届け先がある場合 Snapshot 1`] = `
<div>
  <form>
    <h2>
      お届け先情報の入力
    </h2>
    <fieldset>
      <h3
        style="margin: 0px 0px 8px 0px; padding: 8px 0px 8px 0px; border-bottom: 1px solid #ccc;"
      >
        連絡先
      </h3>
      <div>
        <label>
          電話番号
          <input
            name="phoneNumber"
            type="text"
          />
        </label>
      </div>
      <div>
        <label>
          お名前
          <input
            name="name"
            type="text"
          />
        </label>
      </div>
    </fieldset>
    <fieldset>
      <h3
        style="margin: 0px 0px 8px 0px; padding: 8px 0px 8px 0px; border-bottom: 1px solid #ccc;"
      >
        新しいお届け先を登録しますか？
      </h3>
      <label>
        <input
          name="registerDeliveryAddress"
          type="radio"
          value="0"
        />
        いいえ
      </label>
      <label>
        <input
          name="registerDeliveryAddress"
          type="radio"
          value="1"
        />
        はい
      </label>
    </fieldset>
    <fieldset
      disabled=""
      style="opacity: 0.3;"
    >
      <h3
        style="margin: 0px 0px 8px 0px; padding: 8px 0px 8px 0px; border-bottom: 1px solid #ccc;"
      >
        過去のお届け先
      </h3>
      <select
        name="pastDeliveryAddress"
      >
        <option
          value="address_id_xxxx"
        >
          〒167-0051 東京都杉並区荻窪1-00-00
        </option>
      </select>
    </fieldset>
    <hr />
    <div>
      <button>
        注文内容の確認へ進む
      </button>
    </div>
  </form>
</div>
`;

exports[`過去のお届け先がない場合 Snapshot 1`] = `
<div>
  <form>
    <h2>
      お届け先情報の入力
    </h2>
    <fieldset>
      <h3
        style="margin: 0px 0px 8px 0px; padding: 8px 0px 8px 0px; border-bottom: 1px solid #ccc;"
      >
        連絡先
      </h3>
      <div>
        <label>
          電話番号
          <input
            name="phoneNumber"
            type="text"
          />
        </label>
      </div>
      <div>
        <label>
          お名前
          <input
            name="name"
            type="text"
          />
        </label>
      </div>
    </fieldset>
    <fieldset>
      <h3
        style="margin: 0px 0px 8px 0px; padding: 8px 0px 8px 0px; border-bottom: 1px solid #ccc;"
      >
        お届け先
      </h3>
      <div>
        <label>
          郵便番号
          <input
            name="postalCode"
            placeholder="167-0051"
            type="text"
          />
        </label>
      </div>
      <div>
        <label>
          都道府県
          <input
            name="prefectures"
            placeholder="東京都"
            type="text"
          />
        </label>
      </div>
      <div>
        <label>
          市区町村
          <input
            name="municipalities"
            placeholder="杉並区荻窪1"
            type="text"
          />
        </label>
      </div>
      <div>
        <label>
          番地番号
          <input
            name="streetNumber"
            placeholder="00-00"
            type="text"
          />
        </label>
      </div>
    </fieldset>
    <hr />
    <div>
      <button>
        注文内容の確認へ進む
      </button>
    </div>
  </form>
</div>
`;
